File: backend/ui/src/lib/itemops.ts
// Item ops for the batch screen. Holdings used to be
// reachable only one record at a time, through the editor's item panel;
// `resource: "items"` lets a batch run reshelve a selection through the same
// audited op list, with the same dry-run diff.
import { RESOURCE_ITEMS, type Op, type OpAction } from "./types";

/** The item fields a batch may edit. Mirrors bibframe.ItemFieldNames, and
 *  `TestItemFieldsMatchUI` fails if the two drift.
 *
 *  Barcode is absent on purpose: it names one physical copy, so assigning it
 *  across a selection would mint duplicates. */
export const ITEM_FIELDS: { path: string; label: string }[] = [
  { path: "callNumber", label: "Call number" },
  { path: "location", label: "Shelving location" },
  { path: "note", label: "Item note" },
];

/** Which items an edit reaches. `all` is every copy; `eq` is the ones whose
 *  current value matches exactly (the relocation case: Stacks to Annex leaves
 *  Reference alone); `empty` is the ones missing the field. */
export type ItemGuard = "all" | "eq" | "empty";

/** A field select's option value, encoding the resource it belongs to. Work
 *  fields and item fields share one picker, so the row has to remember which
 *  kind it chose. */
export function fieldKey(resource: string, path: string): string {
  return `${resource}:${path}`;
}

export function parseFieldKey(key: string): { resource: string; path: string } {
  const at = key.indexOf(":");
  if (at < 0) return { resource: "work", path: key };
  return { resource: key.slice(0, at), path: key.slice(at + 1) };
}

export function isItemField(key: string): boolean {
  return parseFieldKey(key).resource === RESOURCE_ITEMS;
}

/** An item field holds one value, so only these two actions mean anything;
 *  the server refuses add/remove rather than reinterpreting them. */
export const ITEM_ACTIONS: OpAction[] = ["set", "clear"];

/** The guard as the wire sends it: `undefined` for every item, otherwise the
 *  exact current value to match. An absent field reads as "", which is what
 *  makes `empty` a plain guard rather than a special case. */
export function guardValue(guard: ItemGuard, where: string): string | undefined {
  if (guard === "all") return undefined;
  if (guard === "empty") return "";
  return where;
}

/** One item-field row as an op, or null when the row is not yet runnable: a
 *  `set` with no value would clear the field, which is what `clear` is for. */
export function itemOp(path: string, action: OpAction, value: string, guard: ItemGuard, where: string): Op | null {
  if (!path) return null;
  const g = guardValue(guard, where);
  if (action === "clear") {
    return { resource: RESOURCE_ITEMS, path, action, ...(g === undefined ? {} : { where: g }) };
  }
  if (action !== "set" || !value) return null;
  return { resource: RESOURCE_ITEMS, path, action: "set", values: [{ v: value }], ...(g === undefined ? {} : { where: g }) };
}

/** How the run summary describes an item op, so the cataloger reads the reach
 *  of the edit before running it rather than after. */
export function itemOpSummary(op: Op): string {
  const field = ITEM_FIELDS.find((f) => f.path === op.path)?.label ?? op.path;
  const scope = op.where === undefined ? "every item" : op.where === "" ? "items with no " + field.toLowerCase() : `items where ${field.toLowerCase()} is “${op.where}”`;
  if (op.action === "clear") return `clear ${field.toLowerCase()} on ${scope}`;
  return `set ${field.toLowerCase()} to “${op.values?.[0]?.v ?? ""}” on ${scope}`;
}

File: bibframe/itemops.go
package bibframe

import (
	"errors"
	"fmt"
	"sort"
	"strings"

	"github.com/freeeve/libcodex/rdf"
)

// Batch-addressable item fields. SetItems replaces an
// Instance's holdings wholesale, which is right for the item panel and wrong
// for a selection: it re-mints every node and churns every quad. An ItemEdit
// is the surgical shape -- one field, on the items that actually change --
// so a batch run's diff shows the relocation and nothing else.
//
// Barcode is deliberately absent. A barcode names one physical copy, so
// assigning one across a selection would mint duplicates; clearing one across
// a selection would silently unlink the shelf from the catalog.
var itemFieldPreds = map[string]string{
	"callNumber": predShelfMark,
	"location":   predPhysicalLocation,
	"note":       PredItemNote,
}

// ErrNoSuchItemField refuses an edit naming a field items do not have, or one
// (barcode) that is not safe to assign across a selection.
var ErrNoSuchItemField = errors.New("no such item field")

// ItemFieldNames lists the batch-editable item fields, sorted, for the error
// messages and the UI's field picker.
func ItemFieldNames() []string {
	names := make([]string, 0, len(itemFieldPreds))
	for name := range itemFieldPreds {
		names = append(names, name)
	}
	sort.Strings(names)
	return names
}

// ItemEdit is one field change applied to every bf:Item in a grain.
//
// Value is the new value; the empty string clears the field. Where, when set,
// restricts the edit to items whose current value is exactly that string,
// which is what makes "move Stacks to Annex" leave the Reference copies
// alone. An item that does not assert the field at all reads as the empty
// string, so Where of "" means "only the ones missing it".
type ItemEdit struct {
	Field string
	Value string
	Where *string
}

// ItemEditPatch translates an ItemEdit into an editorial patch over grainNQ,
// returning the patch and the number of items it touches.
//
// Items already holding the target value are skipped rather than rewritten:
// an unchanged item has no business appearing in a batch run's diff, and a
// remove-then-add of an identical quad would report churn as work.
func ItemEditPatch(grainNQ []byte, e ItemEdit) (Patch, int, error) {
	pred, ok := itemFieldPreds[e.Field]
	if !ok {
		return Patch{}, 0, fmt.Errorf("bibframe: %w: %q (have %s)", ErrNoSuchItemField, e.Field, strings.Join(ItemFieldNames(), ", "))
	}
	if strings.ContainsAny(e.Value, "\n\r") {
		return Patch{}, 0, fmt.Errorf("bibframe: item %s cannot contain a line break", e.Field)
	}
	ds, err := rdf.ParseNQuads(grainNQ)
	if err != nil {
		return Patch{}, 0, err
	}
	// Item nodes are found by their rdf:type, not by an Instance's item-IRI
	// prefix: a grain edited before the prefix scheme settled still types its
	// items, and an edit that silently skipped those would report success
	// while leaving the shelf wrong.
	items := map[string]bool{}
	for _, q := range ds.Quads {
		if q.S.IsIRI() && q.P.Value == rdfTypeIRI && q.O.IsIRI() && q.O.Value == classItem {
			items[q.S.Value] = true
		}
	}
	held := map[string]rdf.Quad{}
	for _, q := range ds.Quads {
		if items[q.S.Value] && q.P.Value == pred {
			held[q.S.Value] = q
		}
	}
	nodes := make([]string, 0, len(items))
	for node := range items {
		nodes = append(nodes, node)
	}
	sort.Strings(nodes)

	patch := Patch{}
	touched := 0
	for _, node := range nodes {
		cur, has := held[node]
		current := ""
		if has {
			current = cur.O.Value
		}
		if e.Where != nil && current != *e.Where {
			continue
		}
		if current == e.Value {
			continue
		}
		if has {
			patch.Remove = append(patch.Remove, rdf.Quad{S: cur.S, P: cur.P, O: cur.O})
		}
		if e.Value != "" {
			patch.Add = append(patch.Add, rdf.Quad{
				S: rdf.NewIRI(node),
				P: rdf.NewIRI(pred),
				O: rdf.NewLiteral(e.Value, "", ""),
			})
		}
		touched++
	}
	return patch, touched, nil
}
